### PR TITLE
Acquire domain ID & workflow ID lock before creating a new workflow

### PR DESCRIPTION
**Pull request by wxing1292**
* Add getOrCreatCurrentWorkflowExecution for current workflow execution lock
* Start workflow execution / signal with start workflow execution, before creating the workflow, will lock on current workflow execution, i.e. run ID being ""
* Rename getOrCreateWorkflowExecutionWithTimeout -> getOrCreateWorkflowExecution
* Rename getOrCreateWorkflowExecution -> getOrCreateWorkflowExecutionForBackground
* Reformat workflow execution context

### CODE REVIEW

File: service/history/historyCache.go [hunk 1]
```diff
@@ -72,38 +72,65 @@ func newHistoryCache(shard ShardContext) *historyCache {
...
+			c.shard,
+			c.executionManager,
+			c.logger,
+		), func(error) {}, nil
+	}
 
+	key := definition.NewWorkflowIdentifier(domainID, workflowID, runID)
+	workflowCtx, cacheHit := c.Get(key).(workflowExecutionContext)
+	if !cacheHit {
+		c.metricsClient.IncCounter(metrics.HistoryCacheGetOrCreateScope, metrics.CacheMissCounter)
```
Comment (venkat1109) [inline service/history/historyCache.go]: scope name should be different given that this is a different method

File: service/history/historyCache.go [hunk 2]
```diff
@@ -72,38 +72,65 @@ func newHistoryCache(shard ShardContext) *historyCache {
 	}
 }
 
-func (c *historyCache) getOrCreateWorkflowExecution(domainID string,
-	execution workflow.WorkflowExecution) (workflowExecutionContext, releaseWorkflowExecutionFunc, error) {
-	return c.getOrCreateWorkflowExecutionWithTimeout(context.Background(), domainID, execution)
-}
+func (c *historyCache) getOrCreatCurrentWorkflowExecution(
```
Comment (venkat1109) [inline service/history/historyCache.go]: this func is mostly the same as getOrCreateWorkflowExecution() except for a flag and another validation call. Consider factoring out the common pieces into a helper method that both of these  can call.

File: service/history/historyCache.go (hunk 1 above)
Comment (wxing1292) [reply service/history/historyCache.go]: good catch

File: service/history/historyCache.go (hunk 2 above)
Comment (wxing1292) [reply service/history/historyCache.go]: sure

**Pull request merged**